File: solutions/3719. Longest Balanced Subarray I/3719.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestBalanced(vector<int>& nums) {
        unordered_map<int,int> freq;
        int evenD = 0, oddD = 0;
        int l = 0, ans = 0;

        for (int r = 0; r < (int)nums.size(); r++) {
            int x = nums[r];
            int &fx = freq[x];
            fx++;
            if (fx == 1) {
                if (x % 2 == 0) evenD++;
                else oddD++;
            }

            while (l <= r && (evenD > oddD || oddD > evenD)) {
                int y = nums[l++];
                int &fy = freq[y];
                fy--;
                if (fy == 0) {
                    if (y % 2 == 0) evenD--;
                    else oddD--;
                    freq.erase(y);
                }
            }

            if (evenD == oddD) ans = max(ans, r - l + 1);
        }

        return ans;
    }
};

File: solutions/3719. Longest Balanced Subarray I/3719.py
from typing import List
from collections import defaultdict

class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        freq = defaultdict(int)
        evenD = 0
        oddD = 0
        l = 0
        ans = 0

        for r, x in enumerate(nums):
            freq[x] += 1
            if freq[x] == 1:
                if x % 2 == 0:
                    evenD += 1
                else:
                    oddD += 1

            # shrink until we can potentially make evenD == oddD
            # (keep reducing when one side is strictly larger)
            while l <= r and (evenD > oddD or oddD > evenD):
                y = nums[l]
                freq[y] -= 1
                if freq[y] == 0:
                    if y % 2 == 0:
                        evenD -= 1
                    else:
                        oddD -= 1
                l += 1

            if evenD == oddD:
                ans = max(ans, r - l + 1)

        return ans
